feat(storage): support for sideloading SSTs

ltagliamonte-dd opened the pull request:
Dear community,

We're working on leveraging RocksDB sideloading to ship updates to our servers instead of loading data through the Redis interface.

Sideloading in RocksDB is more efficient than direct writes because it bypasses the standard write path, reducing write amplification, CPU overhead, and compaction costs. By ingesting pre-built SST files via IngestExternalFile(), it minimizes WAL writes and MemTable pressure, making bulk inserts faster and more efficient.

Since we don't use replication in our setup, we can ship bulk updates directly to each Kvrocks server and load the data efficiently.

We’d love to contribute our sideloading work back to the community and are excited to share our approach. Looking forward to your thoughts!

**Our approach**:
The current approach introduce a new CMD: `SST LOAD path`, which detects the files to load from a folder, and sideloads the data into RocksDB.

ltagliamonte-dd committed:
Merge remote-tracking branch 'ossfork/unstable' into unstable

ltagliamonte-dd committed:
sideloading

ltagliamonte-dd committed:
Merge branch 'unstable' into sideloading

ltagliamonte-dd committed:
format

PragmaTwice commented on src/cli/main.cc:
```diff
@@ -52,6 +52,18 @@ extern "C" void SignalHandler(int sig) {
   }
 }
 
+extern "C" void SigUsr1lHandler(int sig) {
```
It seems weird that we use a custom signal to trigger the ingest operation. Maybe we can consider to introduce a new command.

git-hulk commented:
@ltagliamonte-dd @PragmaTwice Do you think it's a good idea to provide a command for this purpose? 

For example:

```
LOAD SST <local dir>
```
So that it would be easy to ingest SSTs online.

ltagliamonte commented:
@PragmaTwice @git-hulk I'm not married to the use of a signal.
I can definitely add a new server command.
Are you folks ok with the rest of the proposal?

PragmaTwice commented:
> @ltagliamonte-dd @PragmaTwice Do you think it's a good idea to provide a command for this purpose?
> 
> For example:
> 
> ```
> LOAD SST <local dir>
> ```
> 
> So that it would be easy to ingest SSTs online.

Better to be `SST LOAD`.

git-hulk commented:
> > @ltagliamonte-dd @PragmaTwice Do you think it's a good idea to provide a command for this purpose?
> > For example:
> > ```
> > LOAD SST <local dir>
> > ```
> > So that it would be easy to ingest SSTs online.
> 
> Better to be `SST LOAD`.

I proposed  `LOAD SST` for the sake of a command would be better to start with its behavior.

PragmaTwice commented:
But the command name will be `LOAD` if it is. Maybe `LOADSST` or `SSTLOAD` would be better.

`SST LOAD` is similiar to `RDB LOAD`.

git-hulk commented:
> But the command name will be `LOAD` if it is. Maybe `LOADSST` or `SSTLOAD` would be better.
> 
> `SST LOAD` is similiar to `RDB LOAD`.

Makes sense.

ltagliamonte-dd committed:
make sst sideload a cmd

ltagliamonte-dd committed:
cleanup

ltagliamonte-dd committed:
fmt

ltagliamonte-dd committed:
Merge branch 'unstable' into sideloading

ltagliamonte-dd commented:
@git-hulk @PragmaTwice re-worked as new CMD, thanks for the feedback

PragmaTwice commented on kvrocks.conf:
```diff
@@ -1052,5 +1052,10 @@ rocksdb.partition_filters yes
 # Default: 0
 rocksdb.max_compaction_bytes 0
 
+# RocksDB sideloading move_files
+# Move files to the database directory instead of copying them.
+# Default: yes
+rocksdb.sideloading.move_files yes
```
It can be in command arguments instead of a config option.

ltagliamonte-dd replied:
addessed via c9fd7bf9e16b865482895c6984d08d62ba5d4fc5

PragmaTwice commented on src/storage/storage.cc:
```diff
@@ -769,6 +775,65 @@ rocksdb::Status Storage::FlushScripts(engine::Context &ctx, const rocksdb::Write
+  // if the metadata import fails, the imported data will be deleted by the compaction.
+  rocksdb::Status status;
+  // Process default files with no specific column family
+  if (!default_files.empty()) {
+    status = ingestSST(db_->DefaultColumnFamily(), default_ingest_opts_, default_files);
+    if (!status.ok()) {
+      return status;
+    }
+  }
+  // Process metadata files
+  if (!metadata_files.empty()) {
+    status = ingestSST(GetCFHandle(ColumnFamilyID::Metadata), default_ingest_opts_, metadata_files);
+  }
```
I think we have other column families besides these two.

PragmaTwice commented on src/storage/storage.cc:
```diff
@@ -769,6 +775,65 @@ rocksdb::Status Storage::FlushScripts(engine::Context &ctx, const rocksdb::Write
+  std::vector<std::string> metadata_files;
+
+  // Sort files into appropriate vectors based on filename
+  for (const auto &file : sst_files) {
+    if (file.find("metadata") != std::string::npos) {
+      metadata_files.push_back(file);
+    } else {
+      default_files.push_back(file);
+    }
+  }
```
Seems too ad-hoc here? Why not just use column family name instead of hard-coded string here?

ltagliamonte-dd replied:
@PragmaTwice I see the following CF defined:
```
enum class ColumnFamilyID : uint32_t {
  PrimarySubkey = 0,
  Metadata,
  SecondarySubkey,
  PubSub,
  Propagate,
  Stream,
  Search,
};
```
from a quick look at this we could hardcode `default` and search in the folder for all the files that matches those CF? honestly i don't think all CF make sense, eager to hear your opinion.

git-hulk replied:
@ltagliamonte-dd The SST file writer will record the column family if it's assigned, so I think we could find the column family through the SST table properties when loading.

Please refer to https://github.com/facebook/rocksdb/blob/743a02d6f68811bed8c5f3ce5d7e6d9b6bc18783/table/sst_file_writer.cc#L396

PragmaTwice replied:
You can get CF names here: https://github.com/apache/kvrocks/blob/unstable/src/storage/storage.h#L197

ltagliamonte-dd replied:
addressed this via 0ca11358751c7748ca4a0d5228ec4e620bc609aa
please let me know if you have any other feedbacks

ltagliamonte-dd replied:
@git-hulk I looked at your proposal about inserting the CF name in the sstfile and It can't be done atm.
although the SstFileWriter supports the ColumnFamilyHandle parameter, the only 2 exposed method do not:
https://github.com/facebook/rocksdb/blob/v10.0.1/include/rocksdb/c.h#L2301-L2307
So until this is supported upstream (I can sponsor that effort, and improve this code in a second PR) the only way to identify CF seems to use the file name unfortunately.

PragmaTwice commented on src/storage/storage.cc:
```diff
@@ -769,6 +775,65 @@ rocksdb::Status Storage::FlushScripts(engine::Context &ctx, const rocksdb::Write
   return Write(ctx, options, batch->GetWriteBatch());
 }
 
+rocksdb::Status Storage::IngestSST(const std::string &sst_dir, int* files_loaded) {
```
```suggestion
StatusOr<int> Storage::IngestSST(const std::string &sst_dir) {
```

ltagliamonte-dd replied:
addressed via e7907eb6db91814a55a89e36bf5bb7c0913a6598

caipengbo commented:
I feel that this is not a production-usable feature without considering replication. This scenario is too limiting.

ltagliamonte commented:
> I feel that this is not a production-usable feature without considering replication. This scenario is too limiting.

Thank you for the feedback! I completely understand your concern. Please know that this initial version of SST load doesn't rule out the possibility of adding replication in the future. We're committed to ensuring that nothing we implement is backwards incompatible.

git-hulk commented:
> > I feel that this is not a production-usable feature without considering replication. This scenario is too limiting.
> 
> Thank you for the feedback! I completely understand your concern. Please know that this initial version of SST load doesn't rule out the possibility of adding replication in the future. We're committed to ensuring that nothing we implement is backwards incompatible.

Yes, we could allow this command only if it's a standalone instance without any replicas.

ltagliamonte-dd committed:
unit test initial commit

ltagliamonte-dd committed:
address cluster mode comment

ltagliamonte-dd commented:
> > > I feel that this is not a production-usable feature without considering replication. This scenario is too limiting.
> > 
> > 
> > Thank you for the feedback! I completely understand your concern. Please know that this initial version of SST load doesn't rule out the possibility of adding replication in the future. We're committed to ensuring that nothing we implement is backwards incompatible.
> 
> Yes, we could allow this command only if it's a standalone instance without any replicas.

addressed via 68ef7f821a916a2b33324e41bc78340700f504eb

ltagliamonte-dd committed:
address other comments

ltagliamonte-dd committed:
addess StatusOr<int> comment

ltagliamonte-dd committed:
Merge branch 'unstable' into sideloading

ltagliamonte-dd committed:
fmt

ltagliamonte-dd committed:
more fmt

ltagliamonte-dd committed:
more fmt

ltagliamonte-dd committed:
add unit tests

ltagliamonte-dd committed:
fmt

ltagliamonte-dd commented:
@PragmaTwice @git-hulk I believe i've addressed all your comments, the only open one is the handling of other CF.
I've also added unit tests. Looking forward to your feedbacks!

ltagliamonte-dd committed:
more tests

ltagliamonte-dd committed:
init entry

ltagliamonte-dd committed:
fix ci

ltagliamonte-dd committed:
add missing vars

ltagliamonte-dd committed:
add update test case

ltagliamonte-dd committed:
make go fmt work

ltagliamonte-dd committed:
fix rocky linux

ltagliamonte-dd committed:
move install

ltagliamonte-dd committed:
fix other build

ltagliamonte-dd committed:
add debug

ltagliamonte-dd committed:
more debug

PragmaTwice commented on .github/workflows/kvrocks.yaml:
```diff
@@ -558,6 +558,8 @@ jobs:
           echo "DEBUG START"
           ldconfig -p | grep libstdc++
           gcc -print-file-name=libstdc++.so
+          export LIBRARY_PATH=$LIBRARY_PATH:/lib64
+          export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/lib64
```
Hmmm why?

PragmaTwice commented on x.py:
```diff
@@ -267,9 +267,14 @@ def download_package(bindir: str) -> None:
     env.update({
         "CGO_CFLAGS": f"-I{rocksdb}",
-        "CGO_LDFLAGS": f"-L{rocksdb_lib} -L{zlib} -L{z4lib} -L{snappy_lib}",
+        "CGO_LDFLAGS": f"-L{rocksdb_lib} -L{zlib} -L{z4lib} -L{snappy_lib} -L/lib64 -L/usr/lib64 -lstdc++",
+        "LD_LIBRARY_PATH": f"/lib64:/usr/lib64:{env.get('LD_LIBRARY_PATH', '')}"
```
Could you explain this change? Otherwise we need to revert it.

ltagliamonte replied:
Hello @PragmaTwice please ignore the changes in x.py until I'm done fixing the ci that's failing on my fork.

PragmaTwice commented:
The function itself looks good to me. Could you make the implementation more generic? I think in Kvrocks we want a good, complete and well-thought-out implementation instead of "oh, I wrote some code and it seems to work well in my case."

ltagliamonte commented:
> The function itself looks good to me. Could you make the implementation more generic? I think in Kvrocks we want a good, complete and well-thought-out implementation instead of "oh, I wrote some code and it seems to work well in my case."

Thanks I get it I just would like to get more directions on what do you think it's a better/generic design.
As I said I'm not familiar with other CF since my use case is 100% redis hashes.

aleksraiden committed:
Merge branch 'unstable' into sideloading

ltagliamonte-dd committed:
cleanup

ltagliamonte-dd committed:
link path

ltagliamonte-dd committed:
short wait

git-hulk commented:
@ltagliamonte How could we generate the SST if we would like to use this command? I raise this question because I prefer using the same rocksdb version instead of introducing gorocksdb. So I'm wondering if we should support generating SSTs first by providing a subcommand for Kvrocks.

cc @apache/kvrocks-committers

ltagliamonte-dd committed:
debug

ltagliamonte commented:
> @ltagliamonte How could we generate the SST if we would like to use this command? I raise this question because I prefer using the same rocksdb version instead of introducing gorocksdb. So I'm wondering if we should support generating SSTs first by providing a subcommand for Kvrocks.
> 
> cc @apache/kvrocks-committers

Hello @git-hulk in this codebase the use of the gorocksdb library has been introduced just for the unit tests.
I thought about just dropping few pre generated SST files but I didn't think it was the right approach.
I also think that by looking at the unit test code I've added it's going to be extremely simple for any other programming language to generate correct SST files.

git-hulk commented:
> > @ltagliamonte How could we generate the SST if we would like to use this command? I raise this question because I prefer using the same rocksdb version instead of introducing gorocksdb. So I'm wondering if we should support generating SSTs first by providing a subcommand for Kvrocks.
> > cc @apache/kvrocks-committers
> 
> Hello @git-hulk in this codebase the use of the gorocksdb library has been introduced just for the unit tests. I thought about just dropping few pre generated SST files but I didn't think it was the right approach. I also think that by looking at the unit test code I've added it's going to be extremely simple for any other programming language to generate correct SST files.

@ltagliamonte-dd  Yes, I know the gorocksdb is only for test purposes. I see you put a lot of effort into making the CI pass and generating SSTs, which is also a key step for using this command. So I think if we could provide the tool first and reuse this in our test cases.

ltagliamonte commented:
> > > @ltagliamonte How could we generate the SST if we would like to use this command? I raise this question because I prefer using the same rocksdb version instead of introducing gorocksdb. So I'm wondering if we should support generating SSTs first by providing a subcommand for Kvrocks.
> > > cc @apache/kvrocks-committers
> > 
> > 
> > Hello @git-hulk in this codebase the use of the gorocksdb library has been introduced just for the unit tests. I thought about just dropping few pre generated SST files but I didn't think it was the right approach. I also think that by looking at the unit test code I've added it's going to be extremely simple for any other programming language to generate correct SST files.
> 
> @ltagliamonte-dd Yes, I know the gorocksdb is only for test purposes. I see you put a lot of effort into making the CI pass and generating SSTs, which is also a key step for using this command. So I think if we could provide the tool first and reuse this in our test cases.

@git-hulk what kind of tool do you envision? we can think of adding some tooling in a later pr, it's a good idea but it shouldn't block this one wdyt?

git-hulk committed:
Merge branch 'unstable' into sideloading

git-hulk commented:
> > > > @ltagliamonte How could we generate the SST if we would like to use this command? I raise this question because I prefer using the same rocksdb version instead of introducing gorocksdb. So I'm wondering if we should support generating SSTs first by providing a subcommand for Kvrocks.
> > > > cc @apache/kvrocks-committers
> > > 
> > > 
> > > Hello @git-hulk in this codebase the use of the gorocksdb library has been introduced just for the unit tests. I thought about just dropping few pre generated SST files but I didn't think it was the right approach. I also think that by looking at the unit test code I've added it's going to be extremely simple for any other programming language to generate correct SST files.
> > 
> > 
> > @ltagliamonte-dd Yes, I know the gorocksdb is only for test purposes. I see you put a lot of effort into making the CI pass and generating SSTs, which is also a key step for using this command. So I think if we could provide the tool first and reuse this in our test cases.
> 
> @git-hulk what kind of tool do you envision? we can think of adding some tooling in a later pr, it's a good idea but it shouldn't block this one wdyt?

Yes, it won't block this PR. I intend to avoid putting a lot of time into making the CI correct by first building the SSTs tool.

git-hulk commented on src/commands/cmd_server.cc:
```diff
@@ -1370,6 +1370,51 @@ class CommandPollUpdates : public Commander {
+ public:
+  Status Parse(const std::vector<std::string> &args) override {
+    if (args.size() < 3) {
+      return {Status::RedisParseErr, "Invalid number of arguments"};
```
```suggestion
      return {Status::RedisParseErr, errWrongNumOfArguments};
```

git-hulk commented on src/commands/cmd_server.cc:
```diff
@@ -1370,6 +1370,51 @@ class CommandPollUpdates : public Commander {
+    if (args.size() < 3) {
+      return {Status::RedisParseErr, "Invalid number of arguments"};
+    }
+    std::string cmd = util::ToLower(args[1]);
```
Could use `Commandparser` here.

git-hulk commented on src/commands/cmd_server.cc:
```diff
@@ -1410,5 +1455,6 @@ REDIS_REGISTER_COMMANDS(Server, MakeCmdAttr<CommandAuth>("auth", 2, "read-only o
                         MakeCmdAttr<CommandDump>("dump", 2, "read-only", 1, 1, 1),
-                        MakeCmdAttr<CommandPollUpdates>("pollupdates", -2, "read-only admin", NO_KEY), )
+                        MakeCmdAttr<CommandPollUpdates>("pollupdates", -2, "read-only admin", NO_KEY),
+                        MakeCmdAttr<CommandSST>("sst", -3, "write admin", 1, 1, 1), )
```
Need to add the `exclusive` flag for this command.

ltagliamonte-dd committed:
fix script

ltagliamonte-dd committed:
debug

ltagliamonte-dd committed:
missing flag

ltagliamonte-dd committed:
add zlib-devel

aleksraiden committed:
Merge branch 'unstable' into sideloading

ltagliamonte-dd committed:
remove debug

ltagliamonte-dd committed:
set cmake policy

ltagliamonte-dd committed:
move option

ltagliamonte-dd committed:
debug

ltagliamonte-dd committed:
identify asan build

ltagliamonte-dd committed:
revert

ltagliamonte-dd committed:
keep just what's failing

ltagliamonte-dd committed:
tsan lsan

ltagliamonte-dd committed:
fix or

ltagliamonte-dd committed:
add flag

ltagliamonte-dd committed:
disable test for tsan

ltagliamonte-dd committed:
remove flags

ltagliamonte-dd committed:
exclude test when asan

ltagliamonte-dd committed:
fix build

ltagliamonte-dd committed:
add cgo flag when not tsan

ltagliamonte-dd committed:
use same tag

ltagliamonte-dd committed:
debug

ltagliamonte-dd committed:
add ignores

ltagliamonte-dd committed:
ignore

ltagliamonte-dd committed:
exclude

ltagliamonte-dd committed:
add lib to rockylinux 8

ltagliamonte-dd committed:
ignore the others

ltagliamonte-dd committed:
use diff repo

ltagliamonte-dd committed:
add -y

ltagliamonte-dd committed:
get path

ltagliamonte-dd committed:
don't use static

ltagliamonte-dd committed:
add lib path

ltagliamonte-dd committed:
pass gcc path

ltagliamonte-dd committed:
debug

ltagliamonte-dd committed:
remove

ltagliamonte-dd committed:
debug

ltagliamonte-dd committed:
fix path

ltagliamonte-dd committed:
test on rocky 8 and 9

ltagliamonte-dd committed:
make custom env func

ltagliamonte-dd committed:
debug

ltagliamonte-dd committed:
more debug

ltagliamonte-dd committed:
use print

ltagliamonte-dd committed:
match against major only

ltagliamonte-dd committed:
debug

ltagliamonte-dd committed:
more debug

ltagliamonte-dd committed:
debug

ltagliamonte-dd committed:
remove '

ltagliamonte-dd committed:
use str

ltagliamonte-dd committed:
debug

ltagliamonte-dd committed:
remove 32bit

ltagliamonte-dd committed:
add /

ltagliamonte-dd committed:
remove debug

ltagliamonte-dd committed:
debug gcov

ltagliamonte-dd committed:
link gcov

ltagliamonte-dd committed:
run 8

ltagliamonte-dd committed:
print the whole env

ltagliamonte-dd committed:
print env

ltagliamonte-dd committed:
param cov

ltagliamonte-dd committed:
debug

ltagliamonte-dd committed:
full build

ltagliamonte-dd committed:
no debug

ltagliamonte-dd committed:
add back

ltagliamonte-dd committed:
address git-hulk comments

ltagliamonte-dd committed:
loop over all Cf

ltagliamonte-dd committed:
use CommandParser

ltagliamonte-dd commented:
@git-hulk @PragmaTwice i've addressed all comments and the CI is working.
looking forward to any other feedback you may have for me to address.
Thanks for the help in seeing this complete!

git-hulk commented:
CI is blocked by https://github.com/apache/kvrocks/issues/2868

ltagliamonte commented:
> CI is blocked by #2868

FYI Ci run successfully on my end https://github.com/ltagliamonte-dd/kvrocks/actions/runs/14274994555

PragmaTwice committed:
Merge branch 'unstable' into sideloading

PragmaTwice commented on x.py:
```diff
@@ -128,7 +128,7 @@ def build(dir: str, jobs: Optional[int], ghproxy: bool, ninja: bool, unittest: b
     os.makedirs(dir, exist_ok=True)
 
-    cmake_options = ["-DCMAKE_BUILD_TYPE=RelWithDebInfo"]
+    cmake_options = ["-DCMAKE_BUILD_TYPE=RelWithDebInfo", "-DCMAKE_POLICY_VERSION_MINIMUM=3.5"]
```
can be reverted now after #2870.

ltagliamonte-dd replied:
with or without the build seems to fail, is the darwin build flacky?
i've revert as asked.

PragmaTwice commented on x.py:
```diff
@@ -223,6 +223,60 @@ def clang_tidy(dir: str, jobs: Optional[int], clang_tidy_path: str, run_clang_ti
     run(run_command, *options, *regexes, verbose=True, cwd=basedir)
 
 
+def is_rocky_linux():
```
hmmm seems in our CI the rocky linux job works well now.

ltagliamonte-dd replied:
looks like it is needed to do CGO https://github.com/ltagliamonte-dd/kvrocks/actions/runs/14361821853/job/40266547259

ltagliamonte-dd replied:
putting back this change fixes the CI for rocky linux: https://github.com/ltagliamonte-dd/kvrocks/actions/runs/14363316453

ltagliamonte-dd replied:
when we put it [back](https://github.com/ltagliamonte-dd/kvrocks/commit/3fcf93884454d7e74e92937e4337de15e73797a9) CI works for rocky linux https://github.com/ltagliamonte-dd/kvrocks/actions/runs/14363316453

ltagliamonte-dd committed:
revert

ltagliamonte-dd committed:
revert

ltagliamonte-dd committed:
Revert "revert"

This reverts commit 7bebee1ec84af9820cff83043b510e8b8a44b1ee.

ltagliamonte-dd committed:
Revert "revert"

This reverts commit 35ea8e683c38385cc151fb3ab66c07ff824b9593.

ltagliamonte-dd committed:
Reapply "revert"

This reverts commit 37f96a1559ecd78f29568641556117b909136d8a.

ltagliamonte commented:
@PragmaTwice @git-hulk can I please get a final review for this pr?

ltagliamonte-dd committed:
try to fix darwin build

ltagliamonte-dd committed:
add xcode

ltagliamonte-dd committed:
list xcode

ltagliamonte-dd committed:
select xcode

ltagliamonte-dd committed:
Revert "select xcode"

This reverts commit 8b4e630f03313dbe03aa90f5b00235595ebffecd.

ltagliamonte-dd committed:
Revert "list xcode"

This reverts commit 8bcc15c24c1d629277564771beffc94bb0495111.

ltagliamonte-dd committed:
Revert "add xcode"

This reverts commit 2d3ab6d585d2d68f062412ce7f322efa62c97fcf.

ltagliamonte-dd committed:
Revert "try to fix darwin build"

This reverts commit ce541641a4e452c813db504c932a9a4181094d92.

PragmaTwice commented:
Thank you for your effort. All of these build & CI issues will be fixed soon in the seperate PRs, since it is unrelated to the goal of this PR.

ltagliamonte commented:
> Thank you for your effort. All of these build & CI issues will be fixed soon in the seperate PRs, since it is unrelated to the goal of this PR.

Thanks @PragmaTwice.
Is the PR mergeable?

PragmaTwice commented:
> Thanks @PragmaTwice. Is the PR mergeable?

Before that, we need to:
- wait #2876 to be merged to retrigger the CI of this PR;
- revert the changes to `x.py` and `.github/workflows` (if it's really needed, we can open another PR.)

ltagliamonte commented:
@PragmaTwice the changes to x.py and workflows in this PR are needed because of the CGO requirement to compile grocksdb in the unit tests.

PragmaTwice commented:
> @PragmaTwice the changes to x.py and workflows in this PR are needed because of the CGO requirement to compile grocksdb in the unit tests.

Ahh I got it. It's due to this PR brings grocksdb into the golang test. I'll check it later.

ltagliamonte commented:
@PragmaTwice did you had a chance to look at the PR?

ltagliamonte-dd committed:
Merge branch 'apache:unstable' into sideloading

git-hulk committed:
Merge branch 'unstable' into sideloading

ltagliamonte commented:
@PragmaTwice @git-hulk any updates on merging this?

ltagliamonte-dd committed:
Merge branch 'unstable' into sideloading

git-hulk commented on src/commands/cmd_server.cc:
```diff
@@ -1370,6 +1370,49 @@ class CommandPollUpdates : public Commander {
+  Status Parse(const std::vector<std::string> &args) override {
+    if (args.size() < 3) {
+      return {Status::RedisParseErr, errWrongNumOfArguments};
+    }
```
Can remove this check since this command has required at least 3 parameters when building attributes, so it will be checked before executing.
```suggestion
```

git-hulk commented on src/commands/cmd_server.cc:
```diff
@@ -1370,6 +1370,49 @@ class CommandPollUpdates : public Commander {
+  }
+
+  Status Execute([[maybe_unused]] engine::Context &ctx, Server *srv, [[maybe_unused]] Connection *conn,
+                 std::string *output) override {
```
We could check the following conditions and not allow the load:

- The instance is in cluster mode
- The instance has any replicas or is in replica mode.

ltagliamonte-dd replied:
this check is perfomed in https://github.com/ltagliamonte-dd/kvrocks/blob/sideloading/src/storage/storage.cc#L773-L776
since i have already access to _config.

ltagliamonte-dd replied:
@git-hulk how to check `The instance has any replicas or is in replica mode.`?

ltagliamonte-dd replied:
addressed via 66545dada56d37b1090d00e5244723cb6e1ec8c8

git-hulk commented on src/storage/storage.cc:
```diff
@@ -769,6 +770,101 @@ rocksdb::Status Storage::FlushScripts(engine::Context &ctx, const rocksdb::Write
+    }
+    // If no match found, assume it belongs to default CF
+    if (!matched) {
+      cf_files[rocksdb::kDefaultColumnFamilyName].push_back(file);
```
It would be better to return an error if the column family is mismatched.

ltagliamonte-dd replied:
addressed via 04d0aa3dd4e9ff1d64e9b0337c831aeba29528c4

git-hulk commented on .github/workflows/kvrocks.yaml:
```diff
@@ -162,11 +163,13 @@ jobs:
             without_jemalloc: -DDISABLE_JEMALLOC=ON
             with_sanitizer: -DENABLE_ASAN=ON
             compiler: gcc
+            ignore_when_asan: -tags="ignore_when_asan"
```
Where's using this tag?

ltagliamonte-dd replied:
this flag is used here https://github.com/ltagliamonte-dd/kvrocks/blob/sideloading/.github/workflows/kvrocks.yaml#L336

git-hulk commented:
@ltagliamonte-dd I left a few comments, and other looks good to me. To see if @PragmaTwice and @mapleFU have time to take another pass.

ltagliamonte-dd committed:
Update src/commands/cmd_server.cc

Co-authored-by: hulk <hulk.website@gmail.com>

ltagliamonte-dd committed:
return error

ltagliamonte-dd committed:
address hulk comments

ltagliamonte-dd committed:
fix tests

ltagliamonte-dd committed:
Merge branch 'unstable' into sideloading

git-hulk commented on src/commands/cmd_server.cc:
```diff
@@ -1370,6 +1370,52 @@ class CommandPollUpdates : public Commander {
+    if (srv->GetConfig()->cluster_enabled) {
+      return {Status::NotOK, "The SST command is not supported in cluster mode."};
+    }
+    if (srv->IsSlave()) {
```
Need to disallow if it has any replicas.

ltagliamonte-dd replied:
addressed via ce55b4f027f67ebc3bac455f419fd35f0a74ec76

ltagliamonte-dd commented:
@git-hulk is there a method already there I can use to check if the node
has replicas?

On Wed, Apr 16, 2025, 4:40 AM hulk ***@***.***> wrote:

> ***@***.**** approved this pull request.
> ------------------------------
>
> In src/commands/cmd_server.cc
> <https://github.com/apache/kvrocks/pull/2845#discussion_r2046734167>:
>
> > +        } else {
> +          return {Status::RedisParseErr, "movefiles value must be 'yes' or 'no'"};
> +        }
> +      } else {
> +        return {Status::RedisParseErr, "unknown option: " + parser.TakeStr().GetValue()};
> +      }
> +    }
> +    return Commander::Parse(args);
> +  }
> +
> +  Status Execute([[maybe_unused]] engine::Context &ctx, Server *srv, [[maybe_unused]] Connection *conn,
> +                 std::string *output) override {
> +    if (srv->GetConfig()->cluster_enabled) {
> +      return {Status::NotOK, "The SST command is not supported in cluster mode."};
> +    }
> +    if (srv->IsSlave()) {
>
> Need to disallow if it has any replicas.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/kvrocks/pull/2845#pullrequestreview-2772194919>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMKKICBXAG2HIR4377ZWVIL2ZY6SBAVCNFSM6AAAAABZQ7GGRGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDONZSGE4TIOJRHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

git-hulk commented:
> @git-hulk is there a method already there I can use to check if the node has replicas?

You could add a method in `Server` to export the replica count, for example：

```C++

size_t GetReplicaCount() {
   slave_threads_mu_.lock();
   auto replica_count = slave_threads_.size();
   slave_threads_mu_.unlock();
   return replica_count;
} 

```

ltagliamonte-dd committed:
address hulk comment

aleksraiden committed:
Merge branch 'unstable' into sideloading

git-hulk commented:
@ltagliamonte Thanks for your contribution and patience.

ltagliamonte-dd commented:
> @ltagliamonte Thanks for your contribution and patience.

thank you @git-hulk.
Looking forward to pull the nightly!!

caipengbo commented on src/storage/storage.cc:
```diff
@@ -769,6 +770,98 @@ rocksdb::Status Storage::FlushScripts(engine::Context &ctx, const rocksdb::Write
 
+StatusOr<int> Storage::IngestSST(const std::string &sst_dir, const rocksdb::IngestExternalFileOptions &ingest_options) {
+  std::vector<std::string> sst_files;
+  DIR *dir = opendir(sst_dir.c_str());
```
Using the RocksDB API is better:

https://github.com/apache/kvrocks/blob/18e6674eecbfe3361891883f7777c1d188e5ba25/src/storage/storage.cc#L1069

ltagliamonte-dd replied:
Hello @caipengbo the GetChildren does basically the same, put all files in a dir in a vector, why it would be better?

caipengbo replied:
1. RocksDB's API is more portable 
2. The overall code style is more unified

ltagliamonte-dd replied:
addressed via 66eb43e0d356751f3c7b32ea5e4490eb68108a02

caipengbo commented on src/storage/storage.cc:
```diff
@@ -769,6 +770,98 @@ rocksdb::Status Storage::FlushScripts(engine::Context &ctx, const rocksdb::Write
+    return 0;
+  }
+
+  // Create a map to store SST files for each column family
```
I think this comment and some of the comments below are redundant. 

**Good code is comments.**

ltagliamonte-dd replied:
addressed via 66eb43e0d356751f3c7b32ea5e4490eb68108a02

ltagliamonte-dd committed:
address caipengbo comments

ltagliamonte-dd committed:
address caipengbo comments

PragmaTwice commented on x.py:
```diff
@@ -223,6 +223,60 @@ def clang_tidy(dir: str, jobs: Optional[int], clang_tidy_path: str, run_clang_ti
+    if libstdc_folder != "":
+        additional_flags = f"-L{libstdc_folder} -lstdc++"
+
+    with_cov = env.get("WITH_COVERAGE","false")
+    if with_cov == "true":
+        additional_flags += "-lgcov"
```
Do we need to retrieve coverage info for go test cases? Maybe we can just ignore them?

ltagliamonte-dd replied:
this is necessary to support the sonarcloud flag in kvrocks.yaml
```sonarcloud: -DCMAKE_CXX_FLAGS=--coverage```
otherwise build for sonarcloud fails

PragmaTwice commented on src/storage/storage.cc:
```diff
@@ -769,6 +770,90 @@ rocksdb::Status Storage::FlushScripts(engine::Context &ctx, const rocksdb::Write
+  for (const auto &cf : ColumnFamilyConfigs::ListAllColumnFamilies()) {
+    cf_names.emplace_back(cf.Name());
+  }
+  for (const auto &cf_name : cf_names) {
+    cf_files[cf_name] = std::vector<std::string>();
+  }
```
Seems this loop is redundant?

git-hulk replied:
@ltagliamonte-dd You can take a look at this comment. And we could move forward to merging the PR if no more comments.

ltagliamonte-dd replied:
addressed via b2e04e802c979355a76cadd322e06890821b9a60

ltagliamonte-dd committed:
simplify loop

git-hulk commented on src/storage/storage.cc:
```diff
@@ -794,9 +794,7 @@ StatusOr<int> Storage::IngestSST(const std::string &sst_dir, const rocksdb::Inge
-  }
-  for (const auto &cf_name : cf_names) {
-    cf_files[cf_name] = std::vector<std::string>();
+    cf_files[cf.Name()] = std::vector<std::string>();
```
What @PragmaTwice should be is that we don't need to initialize the empty vector for the map?

PragmaTwice replied:
Yeah we can just remove this line.

ltagliamonte replied:
Thanks!
@PragmaTwice can you please approve?

git-hulk commented on src/storage/storage.cc:
```diff
@@ -769,6 +770,88 @@ rocksdb::Status Storage::FlushScripts(engine::Context &ctx, const rocksdb::Write
+  std::vector<std::string> cf_names;
+  for (const auto &cf : ColumnFamilyConfigs::ListAllColumnFamilies()) {
+    cf_names.emplace_back(cf.Name());
+    cf_files[cf.Name()] = std::vector<std::string>();
```
```suggestion
```

git-hulk committed:
Update src/storage/storage.cc

git-hulk commented:
@caipengbo @PragmaTwice To see you guys have further comments, or we could merge after the CI becomes green.

caipengbo reviewed (approved):
LGTM
